### ISSUE TITLE
Qwen 3 conversation with train on inputs false

### ISSUE THREAD

**Issue by arunpatala** [axolotl-ai-cloud/axolotl#3088]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

I am trying to full finetune qwen3 models with a multi turn conversation dataset, and am having trouble with the think tokens during tokenization when using train on inputs=false.

First in multi-turn conversations, when passing messages as the dataset, the empty <think> </think> are only added in the last turn of the assistant message. And I think, when applying masks, the application seems wrong (maybe because of missing think tokens in earlier assistant messages), like below

 <|im_start|>(151644, 151644) user(872, 872)
these tokens should be not trained, but are trained

<|im_start|>(-100, 151644) assistant(-100, 77091)
(-100, 198) {"(-100, 4913) is(-100, 285) Document(-100, 28563) ":(-100, 788) 

after assistant, the tokens should be trained or should be think tokens which need not be trained

As I use VLLM inference, when inferencing the think tokens are added at every turn, so the training should also add at every turn? when using enable_thinking=False.

Is there a better way to specify the tokenizer to add thinking tokens at every turn and also train_on_inputs=False, behave well?

Or should i tokenize myself and add the masking flags and create the dataset?

### Current behaviour

(-100, 198) <|im_start|>(-100, 151644) assistant(-100, 77091)
(-100, 198) {"(-100, 4913) is(-100, 285) ........ <|im_end|>(151645, 151645)
(198, 198) <|im_start|>(151644, 151644) user(872, 872)

(-100, 271) You(-100, 2610)  are(-100, 525)  a(-100, 264)  helpful(-100, 10950)  assistant(-100, 17847) 

### Steps to reproduce

Run Qwen/Qwen3-0.6B with a multi turn converstation dataset with 
train_on_inputs: false
sample_packing: false

axolotl preprocess config.yml --debug

### Config yaml

```yaml
base_model: Qwen/Qwen3-0.6B
auto_resume_from_checkpoints: True


datasets:
  - path: DATA/gpt5/test_merged.jsonl
    type: chat_template
    chat_template: tokenizer_default


test_datasets:
  - path: DATA/gpt5/test_merged.jsonl
    type: chat_template
    split: train
    chat_template: tokenizer_default



evals_per_epoch: 4
saves_per_epoch: 4
output_dir: ./EXP/axolotl/gpt150K_qwen3_V9A

gradient_accumulation_steps: 4
micro_batch_size: 4
num_epochs: 5
warmup_steps: 500
learning_rate: 0.0002
sequence_len: 8192
gradient_checkpointing: true
flash_attention: true
use_tensorboard: true
train_on_inputs: false
sample_packing: false
save_total_limit: 1000
optimizer: adamw_torch_fused
```

### Possible solution

Add a tokenizer flag to add think tokens at every turn of multi turn assistant conversations.
And make sure the mask calculation takes this into consideration.

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.11

### axolotl branch-commit

axolotl, version 0.13.0.dev using docker image:main-latest

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by arunpatala** [comment]
Can provide any other information to reproduce if necessary.

**Comment by arunpatala** [comment]
(-100, 198) <|im_start|>(-100, 151644) assistant(-100, 77091)
(-100, 198) {"(4913, 4913) entries(64751, 64751)  ..............  ]}(13989, 13989) <|im_end|>(151645, 151645)
(-100, 198) <|im_start|>(-100, 151644) user(-100, 872)

this tokenization and masking is working well with Qwen/Qwen2.5-0.5B-Instruct, as it doesnt have think tokens

**Comment by NanoCode012** [comment]
Sorry I missed this issue. I got it confused with another. Could I have an example sample of your dataset to repro it? Feel free to replace with dummy data.

> First in multi-turn conversations, when passing messages as the dataset, the empty are only added in the last turn of the assistant message. 

What does "the empty are only added" mean? 

> And I think, when applying masks, the application seems wrong (maybe because of missing think tokens in earlier assistant messages), like below

Does the template hide think tokens mid convo?

**Comment by NanoCode012** [comment]
Thanks, I can repro the issue with a regular multi-turn dataset (more than 1 assistant). The issue seems similar but slightly different to gpt-oss's.

**Comment by johnr14** [comment]
I am not sure if you mean the same. I implemented this in my train.py, looking how to get it work for axolotl.

```python
def remove_think_tags_except_last(conversation):
    """Clean think tags from non-final assistant turns"""
    logger.debug("Removing think tags except from last assistant turn...")
    assistant_indices = [i for i, turn in enumerate(conversation) if turn.get('role') == 'assistant']

    if not assistant_indices:
        return conversation

    last_assistant_idx = assistant_indices[-1]

    for i in assistant_indices:
        if i != last_assistant_idx:
            content = conversation[i].get('content', '')
            original_content = content
            # Remove think tags using regex to handle nested content
            import re
            content = re.sub(r'<think>.*?</think>', '', content, flags=re.DOTALL)
            conversation[i]['content'] = content
            logger.debug(f"Removed think tags from turn {i}: '{original_content}' -> '{content}'")
        else:
            logger.debug(f"Keeping think tags in last assistant turn {i}: '{conversation[i].get('content', '')}'")

    return conversation
```

```python
def convert_to_turn_by_turn(examples):
    """Split conversations into turn-by-turn format"""
    logger.info("Converting conversations to turn-by-turn format...")
    all_conversations = []

    for conversation in examples[DATASET_COLUMN]:
        accumulated_turns = []
        for i, turn in enumerate(conversation):
            accumulated_turns.append(turn)

            if turn.get('role') == 'assistant':
                processed_turns = []
                for j, acc_turn in enumerate(accumulated_turns):
                    if acc_turn.get('role') == 'assistant':
                        is_last_assistant_in_this_context = True
                        for k in range(j + 1, len(accumulated_turns)):
                            if accumulated_turns[k].get('role') == 'assistant':
                                is_last_assistant_in_this_context = False
                                break

                        content = acc_turn.get('content', '')
                        if not is_last_assistant_in_this_context:
                            import re
                            content = re.sub(r'<think>.*?</think>', '', content, flags=re.DOTALL)
                            content = content.strip()
                        processed_turn = {'role': acc_turn['role'], 'content': content}
                        processed_turns.append(processed_turn)
                    else:
                        processed_turns.append(acc_turn.copy())

                all_conversations.append(processed_turns)

    return {DATASET_COLUMN: all_conversations}
```
But to make it work that way, I split the conversation in sequential turns, with the next one having the previous exchanges but no thinking tokens.

It should make it more similar to real life interaction. However, it generates many more rows to train. My idea was to sample random rows to keep it from exploding in size. So suppose you have a conversation with 15 turns, I split that into 15 different conversations. Then you could sample a random 3 from them. That way you still get to train on thinking traces, without doing it all. I think this is what was used to train [ArliAI](https://huggingface.co/ArliAI/DS-R1-Distill-70B-ArliAI-RpR-v4-Large) but on all generated conversations : 

> Another important thing to get right is to make sure the model is trained on examples that present reasoning blocks in the same way as it encounters it during inference. Which is, never seeing the reasoning blocks in it's context. In order to do this, the training run was completed using axolotl with manual template-free segments dataset in order to make sure that the model is never trained to see the reasoning block in the context. Just like how the model will be used during inference time.

EDIT: Looking back, should prevent splitting from generating a new conversation if there is no thinking tag or empty thinking tag on current assistant turn. Case use: if dataset was generated with Hermes4 that will skip thinking if question was easy. BUT it could be a desired features if <think> is trained to be optional.

**Comment by johnr14** [comment]
Would be nice to have new options for thinking dataset.
```yaml
think_only_on_last_turn: bool # Remove <think>..</think> of precedent turns. Loose other thinking traces.
think_split_conversation_turns : bool | int | ratio # Split one conversation into a series of gradually increasing number of turns with think tags only on the last assistant turn. This will enable training on each thinking trace without seeing other thinking traces in the conversation to represent more real-life reasoning model interaction. Ratio will only apply this on a % of the dataset, int for a max of N samples that gets split.
think_split_conversation_turns_min_turns : int | # Only split conversation that have at least N turns.
think_split_conversation_turns_keep : int | float | none # keep random # or ratio of turns to train on thinking tags. 0.5 = loosing 5 traces on 10. If traces > #, keep at least one trace. None = keep all new conversations.
think_split_conversation_turns_keep_first: bool | float # Always keep first generated conversation + any in think_split_conversation_turns_keep
think_split_conversation_turns_keep_last: bool | float # Always keep last generated conversation + any in think_split_conversation_turns_keep
think_split_conversation_turns_keep_no_think: bool # Should we keep turns where there is no or empty <think> tag. To train optional thinking behavior. 
```
example for doc:
```yaml
think_only_on_last_turn: true # Only train on thinking trace for the last turn. 
think_split_conversation_turns : 0.3 # Split 30% conversations into independent conversations (from those that fill requirements of min_turns).
think_split_conversation_turns_min_turns : 5 # Must have at least 5 turns to have conversation split
think_split_conversation_turns_keep_first: true # Always keep the first turn to be trained on thinking trace.
think_split_conversation_turns_keep_last: 0.2 # force keep last turn 20% of the time
think_split_conversation_turns_keep : 0.1 # from the generated split conversation, keep only an additionnal 10%. Will be cumulative to the 0.2 of think_split_conversation_turns_keep_last and think_split_conversation_turns_keep_first.
```

This should provide good flexibility in training on thinking traces for short or long conversations.

**Comment by NanoCode012** [comment]
I think I get what you kind of mean but I'm not sure whether we want to go into that much detail for thinking only as we usually leave it up to users to preprocess their data. I'll discuss internally on these api options.

> But to make it work that way, I split the conversation in sequential turns, with the next one having the previous exchanges but no thinking tokens.

This was one of the two ideas I am thinking about, to warn users to preprocess their data, so that it's single turn or multi-turn without thinking in between.

The second part is that, we will strip out all <thinking> for non-last assistant turns (which we may have to do either way as it messes with our masking).
